### ISSUE TITLE
Using JSROOT 7 with modern frontend libraries

### ISSUE THREAD

**Issue by 9inpachi** [root-project/jsroot#233]
Hi,

It's great to see that JSROOT now uses ES modules in the code. It seems to work fine in the web browser (with `script[type="module"]`) and also with Node.js but I can't make it work with TypeScript and Angular.

I am trying to update our setup in Phoenix (HSF/phoenix#422) to adapt to the new JSROOT version but the imports are not working correctly.

1. `import { openFile } from 'jsroot/io'` doesn't work because it doesn't recognize the specified exports in `package.json`.
2. If I try to import through the package using `import { openFile } from 'jsroot/modules/io.mjs'`, I get the following error.
    ```
    Error: Module not found: Error: Package path ./mo/modules/io.mjs is not exported from package D:\Repos\phoenix\node_modules\jsroot (see exports field in D:\Repos\phix\oenix\node_modules\jsroot\package.json)
    ```
    I get the same error when trying to compile with webpack.
3. Importing the file directly using `import { openFile } from '../../../../node_modules/jsroot/modules/io.mjs'` gives a lot of errors mainly related to Node.js dependencies like `fs`, `path` etc.

Do you have any idea how to make JSROOT 7 work in a TypeScript and Angular setup?

**Comment by 9inpachi** [comment]
> Did you try simply: import { openFile } from 'jsroot';
By default, jsroot/modules/main.mjs should be imported, which provide all main functions from JSROOT.

I tried that as well. The "module" is set to `main.mjs` but that doesn't get picked up by TypeScript. Most of the packages (that are not ESM) use the "main" field to specify the entry file. I already cloned and tried this but there are errors in the dynamic imports (particularly 'fs', 'path' etc.) in `io.mjs`.

My understanding is that TypeScript uses node for module resolution so it tries to use `fs` and `path` because JSROOT infers the used to be in Node but it's different.

[`phoenix-event-display`](https://github.com/HSF/phoenix/tree/master/packages/phoenix-event-display) is a package itself that can be used by other packages/applications.
[`three.js`](https://github.com/mrdoob/three.js) is also a good reference since it can work with most of the JavaScript module systems.

**Comment by linev** [comment]
> Most of the packages (that are not ESM) use the "main" field to specify the entry file

Most probably, one should use `"main": "./build/jsroot.js"` in package.json.
I just add such entry in master branch. Or you can try modify package.json it locally.

Can you prepare minimal angular project, which only imports jsroot and tries to draw histogram?
And I will try to get it running.

**Comment by 9inpachi** [comment]
> Can you prepare minimal angular project, which only imports jsroot and tries to draw histogram?
And I will try to get it running.

Sure, I will prepare one and share it once I get off of work.

**Comment by linev** [comment]
Here is my modification: https://github.com/linev/jsroot-yarn-webpack/tree/fix

My changes here:  https://github.com/linev/jsroot-yarn-webpack/commit/3b426b1d1fc17fdc8da4e31f7cb6a15c1c298b49

I did minimal changes in jsroot master branch, therefore using directly from github.

Unfortunately, I cannot see any output, but this probably you can debug yourself.

**Comment by 9inpachi** [comment]
I think it doesn't work correctly. Through some quick debugging, I found that instead of importing `./build/jsroot.js`, it imports `./modules/main.mjs`, And at some point in the jsroot code, the execution stops and doesn't continue (no errors shown either). So if I have any code after the jsroot import, it doesn't get executed. I am not sure why that happens.

For example, this code doesn't log anything:

```ts
import * as JSROOT from "jsroot";

console.log("REACHING");
console.log(JSROOT);
```

When I remove the `JSROOT` import, the log statement is reached.

**Comment by linev** [comment]
How I can debug this?

**Comment by 9inpachi** [comment]
After running the application, you can use the Chrome DevTools to add breakpoints to the jsroot code and see how far it reaches.

![image](https://user-images.githubusercontent.com/36920441/173897318-8275f70b-d0da-4679-a809-e561f22b144e.png)

**Comment by 9inpachi** [comment]
I have tried to debug myself but I have no idea why this is happening. The DevTools aren't really helpful here since we don't know at which point the code halts.

I tried to dynamically import jsroot but it seems that the Promise returned from the dynamic `import()` statement never gets resolved/rejected.

```ts
const jsroot = import('jsroot');

jsroot.then(console.log);
console.log(jsroot);
```

**Comment by 9inpachi** [comment]
So after removing the following code, the import works.

`core.mjs`
```js
///_begin_exclude_in_qt5web_
if(isNodeJs()) { node_atob = await import('atob').then(h => h.default); node_xhr2 = await import('xhr2').then(h => h.default); } /// cutNodeJs
///_end_exclude_in_qt5web_
```

`main.mjs`
```js
export * from './hist/TH1Painter.mjs';

export * from './hist/TH2Painter.mjs';

export * from './hist/TH3Painter.mjs';
```

The code that I ran.

```js
import { openFile, draw } from 'jsroot';

async function drawWithJSROOT() {
  let file = await openFile('https://root.cern/js/files/hsimple.root');
  let obj = await file.readObject('hpxpy;1');

  console.log(file, obj);
  draw(document.getElementById('drawing'), obj, 'colz');
}

drawWithJSROOT().finally(() => console.log("HELLO"));
```

The `draw` function doesn't work though but `openFile` does.

I honestly have no idea why this is happening.

**Comment by linev** [comment]
I will investigate it next week

**Comment by linev** [comment]
I add few more changes in JSROOT. 

And create PR for your example: https://github.com/9inpachi/jsroot-minimal/pull/1

This now works for me.

**Comment by 9inpachi** [comment]
@linev, it works now! Thanks.

I see one warning though.

![image](https://user-images.githubusercontent.com/36920441/174966005-d09d7bc4-e19f-4cad-90f4-0fb6c1d8c57f.png)

**Comment by linev** [comment]
Strange, I do not see that warning. Did you apply my PR?

**Comment by linev** [comment]
Now I can reproduce problem.
Conditional check with `process.env?.APP_ENV` inside methods does not work.

But there is other workaround - inserting  ``` /* webpackIgnore: true */ ```

[Found here](https://webpack.js.org/api/module-methods/#magic-comments)

Now it works for me - please try again

**Comment by AlbertoPdRF** [comment]
Hey @linev !

I've also been playing with these changes to improve the ROOT File Viewer extension, bundling it with Webpack, but I can't seem to import the `HierarchyPainter` directly. Any idea why?

**Comment by AlbertoPdRF** [comment]
Yes, it works, I was seeing a different problem.

I don't think I will be able to make it work though, as the `HierarchyPainter` does some direct `document` manipulation, and the `document` isn't available in the Node.js process where the extension is ran :slightly_frowning_face:

**Comment by linev** [comment]
JSROOT should be able to recognize `node.js` environment and avoid usage of `document` instance.

For some tests I also use `HierarchyPainter` in Node - it helps to process in batch complex URLs to produce multiple images as output.

**Comment by AlbertoPdRF** [comment]
Then maybe it isn't recognizing the `node.js` environment because of [this](https://github.com/linev/jsroot-yarn-webpack/blob/master/webpack.config.ts#L63-L66)?

**Comment by linev** [comment]
But this only used with webpack, which should run in normal web browser?

**Comment by AlbertoPdRF** [comment]
Not necessarily, what I was trying to do is to [bundle the extension with webpack](https://code.visualstudio.com/api/working-with-extensions/bundling-extension#using-webpack) as recommended by VS Code so it doesn't include so many files.

**Comment by linev** [comment]
Here how node.js is detected in JSROOT:

https://github.com/root-project/jsroot/blob/master/modules/core.mjs#L19

One can improve it if necessary

**Comment by AlbertoPdRF** [comment]
You can see what I'm trying to do [here](https://github.com/AlbertoPdRF/root-file-viewer/commit/5e4f16e11180b8d5817cff5171b85d7fc7ec814a) -- I can try and walk you through it so you can see the errors I'm getting if you'd like.

**Comment by linev** [comment]
How I can reproduce error you are sees?

**Comment by AlbertoPdRF** [comment]
If you open the base folder in VS Code and install the recommended extensions, you should be able to get the extension compiled and running pressing `F5`. After a new window with the extension running opens, you can open a ROOT File and see the error in the developer console pressing `Ctrl+Shift+I`

**Comment by linev** [comment]
Problem that I do not have `VS Code`. Is if available in community edition?

**Comment by AlbertoPdRF** [comment]
It's free, you can get it from https://code.visualstudio.com/

**Comment by linev** [comment]
@AlbertoPdRF, may be you create new issue?
